src/_ttconv.cpp

```diff
@@ -20,142 +21,70 @@
  */
 class PythonFileWriter : public TTStreamWriter
 {
-    PyObject *_write_method;
+  pybind11::function _write_method;
 
   public:
-    PythonFileWriter()
-    {
-        _write_method = NULL;
-    }
-
-    ~PythonFileWriter()
-    {
-        Py_XDECREF(_write_method);
-    }
-
-    void set(PyObject *write_method)
-    {
-        Py_XDECREF(_write_method);
-        _write_method = write_method;
-        Py_XINCREF(_write_method);
-    }
+    PythonFileWriter(pybind11::object& file_object)
+    : _write_method(file_object.attr("write")) {}
 
     virtual void write(const char *a)
     {
-        PyObject *result = NULL;
-        if (_write_method) {
-            PyObject *decoded = NULL;
-            decoded = PyUnicode_DecodeLatin1(a, strlen(a), "");
-            if (decoded == NULL) {
-                throw py::exception();
-            }
-            result = PyObject_CallFunctionObjArgs(_write_method, decoded, NULL);
-            Py_DECREF(decoded);
-            if (!result) {
-                throw py::exception();
-            }
-            Py_DECREF(result);
+        PyObject* decoded = PyUnicode_DecodeLatin1(a, strlen(a), "");
+        if (decoded == NULL) {
+            throw py::exception();
```

Comment:
Thank you very much for this! There are more suitable reviewers (@ianthomas23 and @anntzer comes to mind), but I think it would be good to replace this exception with "something else" (to get rid of the `py_exceptions.h` import, which we want to get rid of as part of the PyBind11 conversion). I *think* that `std::exception()` is the way to go, possibly with some sensible error message. Although, of course, that general exception is not good etc, but at least it will be consistent with the current behavior.

(This exception is not tested in the original code either.)

Reply:
This should be py::error_already_set (https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-exceptions-from-python-in-c).  We should probably switch the whole thing to write to binary streams (and do something like fh.flush(); convert_ttf_to_ps(..., fh.buffer, ...); fh.buffer.flush() on the Python calling side), but that can be another PR.

Reply:
I switched to using `error_already_set`. Note that this PR is keeping the original behavior on `main` which catches this exception and ignores it:

https://github.com/matplotlib/matplotlib/blob/de8e9d01b0d5bb19c9916f1b49b7eb207899b09a/src/_ttconv.cpp#L139-L142

Reply:
> catches this exception and ignores it
No, returning NULL is the way for C-API functions to signal that an exception has been set and should be propagated to python-land...

Reply:
Ah I see. I misinterpreted the original implementation. Thank you for the explanation!